File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters/Session/ChangeSet.cs
// <copyright file="ChangeSet.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Collections;
    using Meta;
    using Shared.Ranges;

    public sealed class ChangeSet : IChangeSet
    {
        public ChangeSet(Session session, ISet<IStrategy> created, ISet<IStrategy> instantiated)
        {
            this.Session = session;
            this.Created = created ?? EmptySet<IStrategy>.Instance;
            this.Instantiated = instantiated ?? EmptySet<IStrategy>.Instance;
            this.AssociationsByRoleType = new Dictionary<RoleType, ISet<IStrategy>>();
            this.RolesByAssociationType = new Dictionary<AssociationType, ISet<IStrategy>>();
        }

        private Session Session { get; }

        ISession IChangeSet.Session => this.Session;

        public ISet<IStrategy> Created { get; }

        public ISet<IStrategy> Instantiated { get; }

        public IDictionary<RoleType, ISet<IStrategy>> AssociationsByRoleType { get; }

        public IDictionary<AssociationType, ISet<IStrategy>> RolesByAssociationType { get; }

        public void AddSessionStateChanges(IDictionary<IPropertyType, IDictionary<Strategy, object>> sessionStateChangeSet)
        {
            foreach (var kvp in sessionStateChangeSet)
            {
                var strategies = new HashSet<IStrategy>(kvp.Value.Keys);

                switch (kvp.Key)
                {
                    case AssociationType associationType:
                        this.RolesByAssociationType.Add(associationType, strategies);
                        break;
                    case RoleType roleType:
                        this.AssociationsByRoleType.Add(roleType, strategies);
                        break;
                    default:
                        throw new ArgumentOutOfRangeException($"PropertyType {kvp.Key}");
                }
            }
        }

        public void DiffUnit(Strategy association, RelationType relationType, object current, object previous)
        {
            if (!Equals(current, previous))
            {
                this.AddAssociation(relationType, association);
            }
        }

        public void DiffComposite(Strategy association, RelationType relationType, Strategy current, long? previous)
        {
            if (Equals(current?.Id, previous))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, this.Session.GetStrategy((long)previous));
            }

            if (current != null)
            {
                this.AddRole(relationType, current);
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposite(Strategy association, RelationType relationType, long? current, long? previous)
        {
            if (Equals(current, previous))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, this.Session.GetStrategy((long)previous));
            }

            if (current != null)
            {
                this.AddRole(relationType, this.Session.GetStrategy((long)current));
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposite(Strategy association, RelationType relationType, Strategy current, Strategy previous)
        {
            if (Equals(current, previous))
            {
                return;
            }

            if (previous != null)
            {
                this.AddRole(relationType, previous);
            }

            if (current != null)
            {
                this.AddRole(relationType, current);
            }

            this.AddAssociation(relationType, association);
        }

        public void DiffComposites(Strategy association, RelationType relationType, RefRange<Strategy> current, ValueRange<long> previousRange)
        {
            var previous = RefRange<Strategy>.Load(previousRange.Select(v => this.Session.GetStrategy(v)));
            this.DiffComposites(association, relationType, current, previous);
        }

        public void DiffComposites(Strategy association, RelationType relationType, ValueRange<long> currentRange, ValueRange<long> previous)
        {
            var current = RefRange<Strategy>.Load(currentRange.Select(v => this.Session.GetStrategy(v)));
            this.DiffComposites(association, relationType, current, previous);
        }

        public void DiffComposites(Strategy association, RelationType relationType, RefRange<Strategy> current, RefRange<Strategy> previous)
        {
            var hasChange = false;

            foreach (var v in current.Except(previous))
            {
                this.AddRole(relationType, v);
                hasChange = true;
            }

            foreach (var v in previous.Except(current))
            {
                this.AddRole(relationType, v);
                hasChange = true;
            }

            if (hasChange)
            {
                this.AddAssociation(relationType, association);
            }
        }

        private void AddAssociation(RelationType relationType, Strategy association)
        {
            var roleType = relationType.RoleType;

            if (!this.AssociationsByRoleType.TryGetValue(roleType, out var associations))
            {
                associations = new HashSet<IStrategy>();
                this.AssociationsByRoleType.Add(roleType, associations);
            }

            associations.Add(association);
        }

        private void AddRole(RelationType relationType, Strategy role)
        {
            var associationType = relationType.AssociationType;

            if (!this.RolesByAssociationType.TryGetValue(associationType, out var roles))
            {
                roles = new HashSet<IStrategy>();
                this.RolesByAssociationType.Add(associationType, roles);
            }

            roles.Add(role);
        }
    }
}

File: typescript/workspace/system/adapters/src/lib/session/change-set.ts
import { IChangeSet, IObject } from '@allors/workspace-system-domain';
import {
  AssociationType,
  RelationEndType,
  RoleType,
} from '@allors/workspace-system-meta';

import { MapMap } from '../collections/map-map';
import { frozenEmptySet } from '../collections/frozen-empty-set';
import { IRange, Ranges } from '../collections/ranges/ranges';

import { Session } from './session';

export class ChangeSet implements IChangeSet {
  associationsByRoleType: Map<RoleType, Set<IObject>>;
  rolesByAssociationType: Map<AssociationType, Set<IObject>>;

  private ranges: Ranges<IObject>;

  public constructor(public session: Session, public created: Set<IObject>) {
    this.associationsByRoleType = new Map();
    this.rolesByAssociationType = new Map();
    this.created ??= frozenEmptySet as Set<IObject>;

    this.ranges = this.session.ranges;
  }

  public addSessionStateChanges(
    sessionStateChangeSet: MapMap<RelationEndType, IObject, unknown>
  ) {
    for (const [relationEndType, map] of sessionStateChangeSet.mapMap) {
      const strategies = new Set<IObject>();

      for (const [strategy] of map) {
        strategies.add(strategy);
      }

      if (relationEndType.isAssociationType) {
        this.rolesByAssociationType.set(
          relationEndType as AssociationType,
          strategies
        );
      } else if (relationEndType.isRoleType) {
        this.associationsByRoleType.set(
          relationEndType as RoleType,
          strategies
        );
      } else {
        throw new Error(
          `RelationEndType ${relationEndType.name} is not supported`
        );
      }
    }
  }

  public diffUnit(
    association: IObject,
    roleType: RoleType,
    current: unknown,
    previous: unknown
  ) {
    if (current !== previous) {
      this.addAssociationByRoleType(roleType, association);
    }
  }

  public diffCompositeObjectRecord(
    association: IObject,
    roleType: RoleType,
    currentRole: IObject,
    previousRoleId: number
  ) {
    if (currentRole?.id === previousRoleId) {
      return;
    }

    if (previousRoleId != null) {
      const previousRole = (this.session as Session).getObject(previousRoleId);
      if (previousRole) {
        this.addRoleByAssociationType(
          roleType.associationType,
          previousRole
        );
      }
    }

    if (currentRole != null) {
      this.addRoleByAssociationType(roleType.associationType, currentRole);
    }

    this.addAssociationByRoleType(roleType, association);
  }

  public diffCompositeRecordRecord(
    association: IObject,
    roleType: RoleType,
    currentRoleId: number,
    previousRoleId: number
  ) {
    if (currentRoleId === previousRoleId) {
      return;
    }

    if (previousRoleId != null) {
      this.addRoleByAssociationType(
        roleType.associationType,
        (this.session as Session).getObject(previousRoleId)
      );
    }

    if (currentRoleId != null) {
      this.addRoleByAssociationType(
        roleType.associationType,
        (this.session as Session).getObject(currentRoleId)
      );
    }

    this.addAssociationByRoleType(roleType, association);
  }

  public diffCompositeObjectObject(
    association: IObject,
    roleType: RoleType,
    currentRole: IObject,
    previousRole: IObject
  ) {
    if (currentRole === previousRole) {
      return;
    }

    if (previousRole != null) {
      this.addRoleByAssociationType(roleType.associationType, previousRole);
    }

    if (currentRole != null) {
      this.addRoleByAssociationType(roleType.associationType, currentRole);
    }

    this.addAssociationByRoleType(roleType, association);
  }

  public diffCompositeObjectsRecords(
    association: IObject,
    roleType: RoleType,
    currentRoles: IRange<IObject>,
    previousRoleIds: IRange<number>
  ) {
    const previousRoles: IRange<IObject> = previousRoleIds?.map((v) =>
      (this.session as Session).getObject(v)
    );
    this.diffCompositeObjectsObjects(
      association,
      roleType,
      currentRoles,
      previousRoles
    );
  }

  public diffCompositeRecordsRecords(
    association: IObject,
    roleType: RoleType,
    currentRoleIds: IRange<number>,
    previousRoleIds: IRange<number>
  ) {
    const currentRoles: IRange<IObject> = currentRoleIds?.map((v) =>
      (this.session as Session).getObject(v)
    );
    const previousRoles: IRange<IObject> = previousRoleIds?.map((v) =>
      (this.session as Session).getObject(v)
    );
    this.diffCompositeObjectsObjects(
      association,
      roleType,
      currentRoles,
      previousRoles
    );
  }

  public diffCompositeObjectsObjects(
    association: IObject,
    roleType: RoleType,
    currentRoles: IRange<IObject>,
    previousRoles: IRange<IObject>
  ) {
    let hasChange = false;

    for (const v of this.ranges.enumerate(previousRoles)) {
      if (!this.ranges.has(currentRoles, v)) {
        this.addRoleByAssociationType(roleType.associationType, v);
        hasChange = true;
      }
    }

    for (const v of this.ranges.enumerate(currentRoles)) {
      if (!this.ranges.has(previousRoles, v)) {
        this.addRoleByAssociationType(roleType.associationType, v);
        hasChange = true;
      }
    }

    if (hasChange) {
      this.addAssociationByRoleType(roleType, association);
    }
  }

  private addAssociationByRoleType(roleType: RoleType, association: IObject) {
    if (association == null) {
      // TODO: Investigate
      return;
    }

    let associations = this.associationsByRoleType.get(roleType);
    if (!associations) {
      associations = new Set();
      this.associationsByRoleType.set(roleType, associations);
    }

    associations.add(association);
  }

  private addRoleByAssociationType(
    associationType: AssociationType,
    role: IObject
  ) {
    if (role == null) {
      // TODO: Investigate
      return;
    }

    let roles = this.rolesByAssociationType.get(associationType);
    if (!roles) {
      roles = new Set();
      this.rolesByAssociationType.set(associationType, roles);
    }

    roles.add(role);
  }
}
